exec/java-exec/src/main/java/org/apache/drill/exec/store/avro/AvroBatchReader.java

```diff
@@ -0,0 +1,368 @@
+      if (!nextLine(rowWriter)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  @Override
+  public void close() {
+    try {
+      reader.close();
+    } catch (IOException e) {
+      logger.warn("Error closing Avro reader: {}", e.getMessage(), e);
+    } finally {
+      reader = null;
+    }
+  }
+
+  @Override
+  public String toString() {
+    long currentPosition = -1L;
+    try {
+      if (reader != null) {
+        currentPosition = reader.tell();
+      }
+    } catch (IOException e) {
+      logger.trace("Unable to obtain Avro reader position: {}", e.getMessage(), e);
+    }
+    return "AvroBatchReader[File=" + filePath
+      + ", Position=" + currentPosition
+      + "]";
+  }
+
+  /**
+   * Initialized Avro data reader based on given file system and file path.
+   * Moves reader to the sync point from where to start reading the data.
+   *
+   * @param fileSplit file split
+   * @param fs file system
+   * @param opUserName name of the user whom to impersonate while reading the data
+   * @param queryUserName name of the user who issues the query
+   * @return Avro file reader
+   */
+  private DataFileReader<GenericContainer> prepareReader(FileSplit fileSplit, FileSystem fs, String opUserName, String queryUserName) {
+    try {
+      UserGroupInformation ugi = ImpersonationUtil.createProxyUgi(opUserName, queryUserName);
+      DataFileReader<GenericContainer> reader = ugi.doAs((PrivilegedExceptionAction<DataFileReader<GenericContainer>>) () ->
+        new DataFileReader<>(new FsInput(fileSplit.getPath(), fs.getConf()), new GenericDatumReader<GenericContainer>()));
+
+      // move to sync point from where to read the file
+      reader.sync(fileSplit.getStart());
+      return reader;
+    } catch (IOException | InterruptedException e) {
+      throw UserException.dataReadError(e)
+        .message("Error preparing Avro reader")
+        .addContext("Reader", this)
+        .build(logger);
+    }
+  }
+
+  private boolean nextLine(RowSetLoader rowWriter) {
+    try {
+      if (!reader.hasNext() || reader.pastSync(endPosition)) {
+        return false;
+      }
+      container = reader.next(container);
+    } catch (IOException e) {
+      throw UserException.dataReadError(e)
+        .addContext("Reader", this)
+        .build(logger);
+    }
+
+    Schema schema = container.getSchema();
+    GenericRecord record = (GenericRecord) container;
+
+    if (Schema.Type.RECORD != schema.getType()) {
+      throw UserException.dataReadError()
+        .message("Root object must be record type. Found: %s", schema.getType())
+        .addContext("Reader", this)
+        .build(logger);
+    }
+
+    rowWriter.start();
+    List<Schema.Field> fields = schema.getFields();
+    for (Schema.Field field : fields) {
+      String fieldName = field.name();
+      Object value = record.get(fieldName);
+      ObjectWriter writer = rowWriter.column(fieldName);
+      processRecord(writer, value, field.schema());
+    }
```

Comment:
This code is fine; but perhaps slow. It does two hash map lookups per column: one for Avro, another fir the row writer.

Tuples are often defined to be accessed both by name and index. The rowwriter supports this model. I wonder if we can use indexes with Avro?

Looking at the Avro [GenericData](https://github.com/apache/avro/blob/master/lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java) implementation, there is a `get(int index)` method for indexed access.

But, are the column positions the same on each record? Looking at the implementation of `get(String key) `, it seems that Avro does with the row writer does: looks up the column name in the schema, and uses the schema position for column access. This means that columns must have the same position in each row.

This means we can implement the above something like:

```
for (int i = o; i < rowWriter.size(); i++) {
  processRecord(rowWriter.writer(i), record.get(i), fields.get(i));
}
```

Reply:
Good point, replaces with index access. Thanks.